### PR TITLE
Refactor username input lookup component

### PR DESCRIPTION
**Pull request by notbakaneko**
To be used by #11377

Combined the `checkUsernameExists` into the new lookup as well.

### CODE REVIEW

**Comment by notbakaneko** [comment]
I wonder if having a `ref` to the component and just getting the value out directly would be better than passing the updated value through a callback 🤔

File: resources/js/store/store-supporter-tag.tsx [hunk 1]
```diff
@@ -223,16 +229,15 @@ export default class StoreSupporterTag extends React.Component<Props> {
 
   @action
   private readonly getUser = (username: string) => {
-    this.xhr = $.ajax({
-      data: { username },
-      dataType: 'json',
-      type: 'POST',
-      url: route('users.check-username-exists'),
-    });
+    this.xhr = apiLookup([username]);
```
Comment (nanaya) [inline resources/js/store/store-supporter-tag.tsx:232]: the gift link generated in user menu should be `@` prefixed as it was previously strictly username lookup

File: resources/js/store/store-supporter-tag.tsx [hunk 2]
```diff
@@ -223,16 +229,15 @@ export default class StoreSupporterTag extends React.Component<Props> {
...
 
     this.xhr
-      .done((data) => runInAction(() => {
-        this.user = data;
+      .done((response) => runInAction(() => {
+        if (response.users.length > 0) {
+          this.user = response.users[0];
+        } else {
+          this.user = userNotFoundJson;
```
Comment (nanaya) [inline resources/js/store/store-supporter-tag.tsx:239]: the response isn't an observable so might as well just `response.users[0] ?? userNotFoundJson`?

**Review by nanaya — CHANGES_REQUESTED** [review]

File: resources/js/store/store-supporter-tag.tsx (hunk 1 above)
Comment (nanaya) [reply resources/js/store/store-supporter-tag.tsx:232]: or maybe just prefix it before sending the request

File: resources/js/components/username-input.tsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    // spinner should trigger even before request is sent.
+    this.busy = true;
+    this.debouncedLookupUsers();
+
+    if (immediate) {
+      this.debouncedLookupUsers.flush();
+    }
+  }
+
+  private validUsersContains(userIdOrUsername?: string | null) {
```
Comment (nanaya) [inline resources/js/components/username-input.tsx:213]: ...shouldn't this be `Contain` instead? 🤔

**Review by nanaya — CHANGES_REQUESTED** [review]

**Review by nanaya — APPROVED** [review]

**Pull request merged**